libsignal-service/src/account_manager.rs

```diff
@@ -99,34 +99,54 @@ impl<Service: PushService> AccountManager<Service> {
 
         if prekey_count >= PRE_KEY_MINIMUM {
             log::info!("Available keys sufficient");
-            return Ok((pre_keys_offset_id, next_signed_pre_key_id));
+            return Ok((pre_keys_offset_id, signed_pre_key_id));
         }
 
-        let pre_keys = libsignal_protocol::generate_pre_keys(
-            &self.context,
-            pre_keys_offset_id,
-            PRE_KEY_BATCH_SIZE,
-        )?;
-        let identity_key_pair = store_context.identity_key_pair()?;
-        let signed_pre_key = libsignal_protocol::generate_signed_pre_key(
-            &self.context,
-            &identity_key_pair,
-            next_signed_pre_key_id,
-            SystemTime::now(),
-        )?;
-
-        store_context.store_signed_pre_key(&signed_pre_key)?;
-
         let mut pre_key_entities = vec![];
-        for pre_key in pre_keys {
-            store_context.store_pre_key(&pre_key)?;
-            pre_key_entities.push(PreKeyEntity::try_from(pre_key)?);
+        for i in 0..PRE_KEY_BATCH_SIZE {
+            let key_pair = KeyPair::generate(csprng);
+            let pre_key_id =
+                ((pre_keys_offset_id + i) % (PRE_KEY_MEDIUM_MAX_VALUE - 1)) + 1;
+            let pre_key_record = PreKeyRecord::new(pre_key_id, &key_pair);
+            pre_key_store
+                .save_pre_key(pre_key_id, &pre_key_record, None)
+                .await?;
+
+            pre_key_entities.push(PreKeyEntity::try_from(pre_key_record)?);
         }
 
+        // Generate and store the next signed prekey
+        let identity_key_pair =
+            identity_store.get_identity_key_pair(None).await?;
+        let signed_pre_key_pair = KeyPair::generate(csprng);
+        let signed_pre_key_public = signed_pre_key_pair.public_key;
+        let signed_pre_key_signature = identity_key_pair
+            .private_key()
+            .calculate_signature(&signed_pre_key_public.serialize(), csprng)?;
+
+        let unix_time = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap();
+
+        let signed_prekey_record = SignedPreKeyRecord::new(
+            signed_pre_key_id + 1,
+            unix_time.as_secs(),
```

Comment:
Signal seems to like milliseconds for their timestamps, are we sure it's seconds here?

Reply:
Honestly, ::new should accept a DateTime instead :'-)